Audit Logs Rebase

obelisk opened the pull request:
This rebases the audit log system originally written by @CL-Andrew to be on top of all the new code as well as addressing open issues.

This is a move of PR #7428.

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,215 @@
+	select {
+	case l.loggingChannel <- wrappedLog:
+	default:
+		if len(l.loggingChannel) == AUDIT_LOGS_CAPACITY {
+			l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
+		} else {
+			l.logger.Errorw("Could not send log to audit subsystem even though queue has %d space", AUDIT_LOGS_CAPACITY-len(l.loggingChannel))
+		}
```
I think we can simplify this section by dropping the second case since it should never happen by definition.
```suggestion
		l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
```

obelisk replied:
I like having it because it handles all cases and in the event something unexpected happens we will have some indiction that something went wrong.

jmank88 replied:
But the case it is attempting to handle can not happen by definition in the language. IMHO it is a confusing distraction to any readers. :shrug: 

Regardless, as written it could flag false positives with misleading data due to the racy `len` calls. If the channel was not ready to receive we already know that it was full or nil. The only thing to do here is to check for nil, otherwise it was necessarily full.
```suggestion
		if l.loggingChannel != nil {
			l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
		} else {
			l.logger.Errorw("Audit log buffer is nil")
		}
```

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,215 @@
+	"github.com/pkg/errors"
+)
+
+const AUDIT_LOGS_CAPACITY = 2048
```
Can drop the redundant `audit` prefix. 
Does this need to be exported?
Let's use mixed caps too: https://go.dev/doc/effective_go#MixedCaps
Maybe something like this? :shrug:
```suggestion
const bufferCapacity = 2048
```

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,215 @@
+		enabled:         true,
+		serviceURL:      auditLogsURL,
+		serviceHeaders:  headers,
+		jsonWrapperKey:  os.Getenv("AUDIT_LOGS_FORWARDER_JSON_WRAPPER_KEY"),
```
Can we fetch these from a config struct/interface instead of directly from the env vars?

obelisk replied:
Is your preference to modify the GlobalConfig and add these in there?

jmank88 replied:
`GeneralOnlyConfig`

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,233 @@
+	select {
+	case l.loggingChannel <- wrappedLog:
+	default:
+		if len(l.loggingChannel) == bufferCapacity {
```
This `len()` call is racy. There is no way to ensure that it happens atomically with the failure to send to the channel, and it will incorrectly trigger the else case (possibly quite frequently). By this point the goroutine could have been suspended for an arbitrary amount of time and some space could have cleared in the channel concurrently on another thread/routine, yielding a value less than `cap()/buffercapacity`.

But we already know that `len(l.loggingChannel) == cap(l.loggingChannel) == bufferCapacity` at the time we tried to send, because that is the only case other than `nil` for which the channel send would not be ready:
> The default clauses in the select statements execute when no other case is ready, meaning that the selects never block

https://go.dev/doc/effective_go#leaky_buffer

obelisk replied:
Fixed

jmank88 commented on core/logger/logger.go:
```diff
@@ -215,6 +215,8 @@ type Config struct {
 	FileMaxSizeMB  int
 	FileMaxAgeDays int
 	FileMaxBackups int // files
+	Hostname       string
+	ChainlinkDev   bool
```
Are these used? Or leftover from a different plan?

obelisk replied:
Looks like they are left over. Will fix.

jmank88 commented on core/services/chainlink/application.go:
```diff
@@ -198,6 +201,12 @@ func NewApplication(opts ApplicationOpts) (Application, error) {
 	unrestrictedHTTPClient := opts.UnrestrictedHTTPClient
 
+	// Configure and optionally start the audit log forwarder service
+	auditLogger, err := audit.NewAuditLogger(globalLogger, cfg.AuditLogger())
```
Overloading the name `AuditLogger` for both `audit.AuditLogger` and `audit.AuditLoggerConfig` seems a little confusing. What about being a little more explicit?
```suggestion
	auditLogger, err := audit.NewAuditLogger(globalLogger, cfg.GetAuditLoggerConfig())
```
Alternatively, you could use an interface for the config type, and then implement it directly via `cfg` so it can just be passed as-is without any extra indirection:
```suggestion
	auditLogger, err := audit.NewAuditLogger(globalLogger, cfg)
```
FWIW This is a little more aligned with the rest of the system, but nbd.

obelisk replied:
Yeah I like this

samsondav commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,231 @@
+
+	// Start our go routine that will receive logs and send them out to the
+	// configured service.
+	go auditLogger.auditLoggerRoutine()
```
Generally speaking, we don't like to silently spin off goroutines in a `NewXXX` function. Please implement `service.ServiceCtx` interface instead.

samsondav commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,231 @@
+	select {
+	case l.loggingChannel <- wrappedLog:
+	default:
+		if l.loggingChannel == nil {
```
This case seems impossible to reach

samsondav commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,231 @@
+// /
+// / This function calls postLogToLogService which blocks.
+func (l *AuditLoggerService) auditLoggerRoutine() {
+	for event := range l.loggingChannel {
```
Please implement as a `service.ServiceCtx` that responds to `Close`.

For a (relatively) simple example, see `EthResender`.

samsondav commented on core/sessions/orm.go:
```diff
@@ -172,10 +176,12 @@ func (o *orm) CreateSession(sr SessionRequest) (string, error) {
 	// Do email and password check first to prevent extra database look up
 	// for MFA tokens leaking if an account has MFA tokens or not.
 	if !constantTimeEmailCompare(strings.ToLower(sr.Email), strings.ToLower(user.Email)) {
+		o.auditLogger.Audit(context.Background(), audit.AuthLoginFailedEmail, map[string]interface{}{"email": sr.Email})
```
It's probably no longer necessary to pass a context in here since it makes no external requests now.

samsondav commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	// Send up to HEC log collector
+	httpClient := &http.Client{Timeout: time.Second * 10}
```
This should probably be set on the AuditLoggerService and not re-used each time.

You may also consider passing it into the constructor, this would allow you to easily mock it in a test.

jmank88 commented on core/services/chainlink/testdata/config-full.toml:
```diff
@@ -3,6 +3,11 @@ InsecureFastScrypt = true
 
+[AuditLoggerConfig]
+forwardToUrl = "http://localhost:9898"
+headers = []
```
Could we include some dummy values to check that everything parses?

obelisk replied:
Since this is setting the config directly I think it bypasses the parsing process.

jmank88 replied:
New config parses directly from the TOML files. This file is used to sanity check example values for all of the fields.

jmank88 replied:
If the intent is for users to input the encoded format, then the type used in the TOML struct will have to implement [TextMarshaler](https://pkg.go.dev/encoding#TextMarshaler)/[TextUnmarshaler](https://pkg.go.dev/encoding#TextUnmarshaler)

jmank88 commented on core/services/chainlink/testdata/config-full.toml:
```diff
@@ -3,6 +3,11 @@ InsecureFastScrypt = true
 RootDir = 'test/root/dir'
 ShutdownGracePeriod = '10s'
 
+[AuditLoggerConfig]
```
I think we can leave off the `Config` suffix from the field name, since it is implied by the context.
```suggestion
[AuditLogger]
```

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,230 @@
+}
+
+type AuditLoggerConfig struct {
+	forwardToUrl   string
+	environment    string
+	jsonWrapperKey string
+	headers        []serviceHeader
```
For this to be directly compatible with the new toml config, it should use capitalized field names, and pointer types. `models.URL` is helpful too.
```suggestion
	ForwardToUrl   *models.URL
	Environment    *string
	JsonWrapperKey *string
	Headers        []serviceHeader
```

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,230 @@
+}
+
+// Configurable headers to include in POST to log service
+type serviceHeader struct {
+	header string
+	value  string
+}
```
These should be exported:
```suggestion
type ServiceHeader struct {
	Header string
	Value  string
}
```
I don't think that `*string` is necessary in this case.

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,230 @@
+	value  string
+}
+
+type WrappedAuditLog struct {
```
It looks like this type could be unexported?
```suggestion
type wrappedAuditLog struct {
```

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,230 @@
+
+	// Send up to HEC log collector
+	httpClient := &http.Client{Timeout: time.Second * 10}
+	req, _ := http.NewRequest("POST", l.serviceURL, bytes.NewReader(serializedLog))
```
Unhandled err?

obelisk replied:
Good catch

jmank88 commented on core/web/jobs_controller.go:
```diff
@@ -156,6 +158,9 @@ func (jc *JobsController) Create(c *gin.Context) {
 		return
 	}
 
+	jbj, _ := json.Marshal(jb)
```
Unhandled error?

obelisk commented on core/sessions/orm.go:
```diff
@@ -234,6 +241,14 @@ func (o *orm) CreateSession(sr SessionRequest) (string, error) {
+	uwasj, err := json.Marshal(uwas)
+	if err != nil {
+		lggr.Errorf("error in Marshal credentials: %s", err)
+		return "", err
```
@CL-Andrew is it your intention to break logging in if marshalling fails?

jmank88 commented on core/config/v2/types.go:
```diff
@@ -19,6 +20,7 @@ import (
 // Core holds the core configuration. See chainlink.Config for more information.
 type Core struct {
 	// General/misc
+	AuditLoggerConfig   *audit.AuditLoggerConfig
```
```suggestion
	AuditLogger   *audit.AuditLoggerConfig
```

jmank88 commented on core/sessions/orm.go:
```diff
@@ -234,6 +241,13 @@ func (o *orm) CreateSession(sr SessionRequest) (string, error) {
+	uwasj, err := json.Marshal(uwas)
+	if err != nil {
+		lggr.Errorf("error in Marshal credentials: %s", err)
+	}
+	o.auditLogger.Audit(audit.AuthLoginSuccessWith2FA, map[string]interface{}{"email": sr.Email, "credential": string(uwasj)})
```
Should/must this be skipped if we fail to marshal?
```suggestion
	} else {
		o.auditLogger.Audit(audit.AuthLoginSuccessWith2FA, map[string]interface{}{"email": sr.Email, "credential": string(uwasj)})
	}
```

jmank88 commented on core/config/v2/types.go:
```diff
@@ -19,6 +20,7 @@ import (
 // Core holds the core configuration. See chainlink.Config for more information.
 type Core struct {
 	// General/misc
+	AuditLogger         *audit.AuditLoggerConfig
```
Let's move this down below with the other toml 'table's. There's not really a strict order among them though, so wherever you think makes sense :shrug:

obelisk replied:
I guess I was going for alphabetical order and not mixing any more. I don't feel strongly if you want it moved though

jmank88 replied:
Well the tables/structs are below this set of `General/misc` fields since that is how the toml lib encodes it. This one could pair with the `Log` table. :shrug:

jmank88 commented on core/web/nodes_controller.go:
```diff
@@ -25,6 +27,8 @@ type nodesController[I chains.ID, N chains.Node, R jsonapi.EntityNamer] struct {
 	errNotEnabled error
 	newResource   func(N) R
 	createNode    func(*gin.Context) (N, error)
+	lggr          logger.Logger
```
Unused?
```suggestion
	lggr          logger.Logger
```

jmank88 commented on core/web/keys_controller.go:
```diff
@@ -67,6 +73,21 @@ func (kc *keysController[K, R]) Create(c *gin.Context) {
+
+	// Emit audit log, determine if Terra or Solana key
+	switch unwrappedKey := any(key).(type) {
+	case terrakey.Key:
```
Ah, we still need to find a different way to handle these different cases, since enumerating them internally defeats the purpose of the generic code. It looks like interfaces should be able to do most of the work for the data, if we pass in just the set of event ids instead. I can help if the generics are confusing.

obelisk replied:
I think we can still handle this in a follow up. I believe @CL-Andrew has a shortcut task for it.

jmank88 replied:
I disagree. This is a blocker. It is already broken, and doesn't show until runtime.

jmank88 commented on core/web/chains_controller.go:
```diff
@@ -101,6 +108,12 @@ func (cc *chainsController[I, C, R]) Create(c *gin.Context) {
 
+	chainj, err := json.Marshal(chain)
+	if err != nil {
+		cc.lggr.Errorf("Unable to marshal chain to json", "err", err)
+	}
+	cc.auditLogger.Audit(audit.ChainAdded, map[string]interface{}{"chain": chainj})
```
If an error is returned, there is no guarantee about the contents of `chainj`. It might be empty, but could also contain partial or non-sensical data. It could be set nil explicity:
```suggestion
		cc.lggr.Errorf("Unable to marshal chain to json", "err", err)
		chainj = nil
	}
	cc.auditLogger.Audit(audit.ChainAdded, map[string]interface{}{"chain": chainj})
```
Or we could log something different that notes the error? :shrug:

jmank88 commented on core/services/chainlink/application.go:
```diff
@@ -198,6 +201,14 @@ func NewApplication(opts ApplicationOpts) (Application, error) {
+	// Configure and optionally start the audit log forwarder service
+	auditLogger, err := audit.NewAuditLogger(globalLogger, cfg.AuditLoggerConfig())
+	if err != nil {
+		globalLogger.Info("Audit logger could not be started. err: ", err)
```
Should this be a fatal error? Generally if a service cannot be created (let alone started) then something is very wrong that needs to be fixed. I think it would make sense to return a wrapped error here, like pyroscope and nurse do below.

obelisk replied:
This should not be a fatal error because it also occurs when there is no configuration which is probably most of the time.

To stay generic, the AuditLogger takes a config and errors if it cannot create one successfully, in this case though making it a fatal error would mean the logger needs to always be configured

jmank88 replied:
Oh, that is surprising and non-standard for both Go and our own internal style. We would prefer that errors stay reserved for true error conditions. Normally we would first do a config check to see if something is enabled before attempting to create it.

jmank88 replied:
How about something like this?
```go
if cfg.AuditLoggerEnabled() {
	auditLogger, err = audit.NewAuditLogger(globalLogger, cfg.AuditLoggerConfig())
	if err != nil {
		return nil, errors.Wrap(err, "failed to start Audit Logger")
	}
} else {
	globalLogger.Info("Audit Logger is disabled")
	auditLogger = audit.NoopAuditLogger()
}
```

obelisk replied:
This is pretty much the new system but instead of implementing a NoopAuditLogger which would also need to implement service.ServiceCtx we return an empty AuditLoggerService{} which also does nothing.

jmank88 replied:
> instead of implementing a NoopAuditLogger which would also need to implement service.ServiceCtx 

I am not proposing this.

You can still use a zero value `AugitLoggerService`. But for readability, forward compatibility, deduplication, etc. you should define a func to return it. Especially for the callers outside the package or in tests.
```go
package audit

func NoopAuditLogger() { return &AuditLoggerService{} }
```
In fact, `AuditLoggerService` should be unexported as `auditLoggerService` if possible.

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,312 @@
+// Parses and validates the AUDIT_LOGS_* environment values and returns an enabled
+// AuditLogger instance. If the environment variables are not set, the logger
+// is disabled and short circuits execution via enabled flag.
+func NewAuditLogger(logger logger.Logger, config *AuditLoggerConfig) (AuditLogger, error) {
+	if config == nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - no configuration")
+	}
```
Is this a feature or a quirk? Can we just use the struct type directly instead?
```suggestion
func NewAuditLogger(logger logger.Logger, config AuditLoggerConfig) (AuditLogger, error) {
```

obelisk replied:
AuditLoggerConfig can be nil

obelisk replied:
This is because when configuration fails (which is a normal thing and happens when there is no configuration) the pointer to this struct is nil

jmank88 replied:
> when configuration fails (which is a normal thing and happens when there is no configuration) 

IMHO this is needlessly confusing. We should follow standard idiomatic patterns from Go, the application, and the surrounding code. Is there some reason this system needs to break form that I am overlooking?

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,312 @@
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - unable to get hostname: %s", err)
```
This is pretty unusual. IMO using the idiomatic error _or_ result pattern will makes things much more readable at the callsite.
```suggestion
		return nil, errors.Errorf("Audit Log initialization error - unable to get hostname: %s", err)
```

obelisk replied:
Currently we use this to avoid code duplication implementing a null logger.

jmank88 replied:
This is a non-idiomatic, confusing way to implement a null logger in Go and within this code base. Please follow the traditional convention of the result being unused when error is present. The caller can explicitly create a null logger if they still need one.

samsondav commented on core/cmd/client.go:
```diff
@@ -217,6 +218,12 @@ func (n ChainlinkAppFactory) NewApplication(ctx context.Context, cfg config.Gene
+	// Configure and optionally start the audit log forwarder service
+	auditLogger, err := audit.NewAuditLogger(appLggr, cfg.AuditLoggerConfig())
+	if err != nil {
+		appLggr.Info("Audit logger could not be started. err: ", err)
```
```suggestion
		appLggr.Infow("Audit logger could not be started", "err", err)
```

Can you check other uses of logger to make sure they output structured logging correctly?

samsondav replied:
Actually, I would recommend failing application boot if this thing fails to start.

samsondav replied:
It also needs to be closed on application exit.

obelisk replied:
Since this is a subservice will the that be handled automatically during shutdown where close is called? The requests now use context.Context so should be killed when shutdown is requested.

jmank88 replied:
Yes it will be started/closed as part of the slice variable named `subservices`, but it is misnamed: https://github.com/smartcontractkit/chainlink/pull/7487

samsondav commented on core/config/general_config.go:
```diff
@@ -484,6 +486,32 @@ func (c *generalConfig) AppID() uuid.UUID {
+
+	if err != nil {
+		c.lggr.Errorf("Audit logger configuration failed with: %s", err)
+		return nil
```
Can this cause panics elsewhere if they are expecting a non-nil configuration?

obelisk replied:
This is only used when building the AuditLogger which also take a pointer and checks if it's nil

jmank88 replied:
Quietly disabling a system that is accidentally misconfigured is an anti-pattern that we have been trying to avoid.

A lot of this could be avoided if the individual config options had their own methods like other systems (with errors when appropriate). Then the constructor could handle things directly instead of having this delayed/deffered indirection.

jmank88 replied:
Or instead of that refactor ^ why not just return an error here also? That seems like most of the same benefits without the mess.

obelisk replied:
How do we want to handle the difference between:
* No configuration - A valid state
* Bad configuration - An invalid state

In Rust parlance I would use the type `Result<Option<AuditLoggerConfig>, ConfigurationError>>` which is what I tried to do here with the pointer forming the option

jmank88 replied:
It would be OK to use nil config for 'disabled':
```go
func foo() (*Config, error) {
  // no config
  return nil, nil

  // bad config
  return nil, errors.Wrapf(...

  // happy path
  return &Config{...}, nil
}
```
But we also commonly have a very explicit interface method for `XXXEnabled() bool`, even if the implementation is derived from the presence of some other var(s), so this signature could be simpler:
```go
func foo() (Config, error) {
  // bad config
  return Config{}, errors.Wrapf(...

  // happy path
  return Config{...}, nil
}
```
Then the caller could check for `Config.Enabled()`, or infer it themselves.

obelisk replied:
No longer using pointers

samsondav commented on core/internal/cltest/cltest.go:
```diff
@@ -332,6 +333,19 @@ func NewApplicationWithConfig(t testing.TB, cfg *configtest.TestGeneralConfig, f
+	}
+
+	if auditLogger == nil {
+		auditLogger = &audit.AuditLoggerService{}
```
You might consider using the null pattern instead here.

obelisk replied:
This will mean checking nil at every audit callsite. Is this preferred?

jmank88 replied:
I think that Sam meant a `NullAuditLogger` type, equivalent to a `NoopLogger`. If you use a Null/Noop func as mentioned elsewhere, then you can fiddle with the implementation without having to refactor all of these callers.

samsondav commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,312 @@
+
+func NewAuditLoggerConfig(forwardToUrl string, isDev bool, jsonWrapperKey string, encodedHeaders string) (AuditLoggerConfig, error) {
+	if forwardToUrl == "" {
+		return AuditLoggerConfig{}, errors.Errorf("No forwardToURL provided")
```
No point in using `errors.Errorf` if there is nothing to interpolate.
```suggestion
		return AuditLoggerConfig{}, errors.New("No forwardToURL provided")
```

samsondav commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,312 @@
+	}
+
+	if _, err := models.ParseURL(forwardToUrl); err != nil {
+		return AuditLoggerConfig{}, errors.Errorf("forwardToURL value is not a valid URL")
```
Same here. Please check for other uses.

samsondav commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,312 @@
+	for _, header := range l.headers {
+		req.Header.Add(header.Header, header.Value)
+	}
+	resp, err := l.loggingClient.Do(req)
```
This should take the context, so that any in-flight requests are cancelled immediately on application shutdown.

jmank88 replied:
I think it will need to happen on 268 via https://pkg.go.dev/net/http#NewRequestWithContext

obelisk replied:
Done

samsondav commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,312 @@
+	for {
+		select {
+		case <-l.ctx.Done():
+			// I've made this an error since we expect it should never happen.
```
It will happen when the AuditLoggerService is `Close`d.

samsondav commented on core/sessions/orm.go:
```diff
@@ -234,6 +241,14 @@ func (o *orm) CreateSession(sr SessionRequest) (string, error) {
+	if err != nil {
+		lggr.Errorf("error in Marshal credentials: %s", err)
+	} else {
+		o.auditLogger.Audit(audit.AuthLoginSuccessWith2FA, map[string]interface{}{"email": sr.Email, "credential": string(uwasj)})
```
Is this sensitive info?

samsondav commented on core/sessions/reaper_test.go:
```diff
@@ -33,7 +34,7 @@ func TestSessionReaper_ReapSessions(t *testing.T) {
 	lggr := logger.TestLogger(t)
 	cfg := cltest.NewTestGeneralConfig(t)
-	orm := sessions.NewORM(db, config.SessionTimeout().Duration(), lggr, cfg)
+	orm := sessions.NewORM(db, config.SessionTimeout().Duration(), lggr, cfg, &audit.AuditLoggerService{})
```
This is the same idea as the NullAuditLogger I was suggesting above.

samsondav commented on docs/CHANGELOG.md:
```diff
@@ -21,6 +21,45 @@ as an estimated gas limit (up to `ETH_GAS_LIMIT_MAX`, with `1,000,000,000` defau
+
+    - Default: _none_
+
+    When set, this environment variable configures and enables an optional HTTP logger which is used specifcally to send audit log events. Audit logs events are emitted when specific actions are performed by any of the users through the node's API. The value of this variable should be a full URL. Log items will be sent via POST
```
```suggestion
    When set, this environment variable configures and enables an optional HTTP logger which is used specifically to send audit log events. Audit logs events are emitted when specific actions are performed by any of the users through the node's API. The value of this variable should be a full URL. Log items will be sent via POST
```

samsondav commented on core/logger/audit/audit_logger_test.go:
```diff
@@ -0,0 +1,127 @@
+	auditLoggerService.SetLoggingClient(&mockHTTPClient)
+	assert.NoError(t, auditLoggerService.Ready())
+
+	// Create a new chainlink test application passing in our test logger
+	// and audit logger
+	app := cltest.NewApplication(t, logger, auditLogger)
+	require.NoError(t, app.Start(testutils.Context(t)))
```
This gives more coverage but is probably overkill, I'd suggest testing the audit logger in isolation, and if you want to test actual audit logging points, do that in a separate integration test using a mock audit logger object.

obelisk replied:
I do like the base of having better coverage. We are planning to expand the Audit implementation in the future so having more coverage will be better.

obelisk commented:
I've rewritten the configuration system so we have validation which is still private to the module. I've also removed lots of pointer uses to reduce crashes.

On the point of a Noop/NilLogger an empty struct currently fulfills this role pretty well and doesn't even start a subservice when not configured. Calls into it Audit get dropped easily as well.

jmank88 commented on core/config/v2/types.go:
```diff
@@ -30,7 +31,8 @@ type Core struct {
 	TelemetryIngress *TelemetryIngress
 
-	Log *Log
+	AuditLogger audit.AuditLoggerConfig
```
Does this have to be a pointer to prevent printing zero values in the toml?
```suggestion
	AuditLogger *audit.AuditLoggerConfig
```

obelisk replied:
Does it? I would prefer this not be a pointer to cut down on the number of places it could accidentally dereferenced.

jmank88 replied:
Yes. It must be a pointer to distinguish between omission and empty values when marshaling and unmarshaling.

jmank88 replied:
`nil` is meaningful here, and they get filled in with default values before use. This process is tested for completeness as well.

obelisk replied:
Converting to pointer

jmank88 commented on core/logger/audit/audit_logger.go:
```diff
@@ -0,0 +1,346 @@
+// /
+// / Checks all sorts of things to eliminate nil pointers and invalid endpoints
+// / as well as parsing out the headers into format that's easier to use later.
+func validateConfig(alc AuditLoggerConfig) (validatedAuditLoggerConfig, error) {
```
This is functional for the current legacy configuration, but it is unfortunately no longer compatible with the new TOML config.

jmank88 replied:
It may be simpler to split the types between toml and legacy if that helps, but I really think that most of these problems would go away if you switched to the usual convention of receiving an interface of individual methods. :shrug:

obelisk replied:
What is incompatible about this function?

jmank88 replied:
The new TOML config does validation up front, and per-field via types. Using plain strings and manually validating later breaks this model. The previous version was very close to what we needed.

obelisk replied:
We need to manually validate later or duplicate the validation code because the fields are public

jmank88 replied:
I don't understand what you mean. If we enforce as much as possible via strict types in the config struct/interface then the validation code can be limited to `UnmarshalText` implementations and trivially reused to parse from anywhere else if necessary.

jmank88 replied:
See `models.URL` for example:
https://github.com/smartcontractkit/chainlink/blob/af5b414cb86677a86f20993ded92fcce63ad11ca/core/store/models/common.go#L512

jmank88 replied:
There might still be some duplication for higher level validation, but if so then it's just necessary until we retire legacy config :shrug: 
It's also possible the new config `ValidateConfig()` impl would be reusable, but it has to exist for new config going forward either way.

jmank88 commented on core/services/chainlink/config_general.go:
```diff
@@ -166,6 +167,26 @@ func (g *generalConfig) AllowOrigins() string {
 }
 
+func (g *generalConfig) AuditLoggerEnabled() bool {
+	return *g.c.AuditLogger.Enabled
```
On second thought, this explicit `Enabled *bool` field may not be strictly necessary. :thinking:  I think it would be ok to drop the _field_, but keep the _method_ and implement based on :
```suggestion
	return g.c.AuditLogger != nil
```
But really nbd in for this PR, as long as it is internally consistent. We haven't sorted this out as a whole for new config yet, but will be working on it very soon, and it will be easy to tweak this if necessary.

jmank88 commented on core/config/general_config.go:
```diff
@@ -484,6 +486,37 @@ func (c *generalConfig) AppID() uuid.UUID {
+}
+
+func (c *generalConfig) AuditLoggerEnvironment() string {
+	if c.viper.GetBool(envvar.Name("Dev")) {
```
Is there a `Dev()` method to call instead?
```suggestion
	if c.Dev() {
```

jmank88 commented on core/web/eth_keys_controller.go:
```diff
@@ -288,6 +308,13 @@ func (ekc *ETHKeysController) Export(c *gin.Context) {
+
+	ekc.App.GetAuditLogger().Audit(audit.KeyExported, map[string]interface{}{
+		"type":    "ethereum",
+		"id":      "", // TODO: Figure out what the best way to fetch this is for consistency
```
Still todo?

jmank88 commented on core/web/keys_controller.go:
```diff
@@ -100,6 +133,20 @@ func (kc *keysController[K, R]) Import(c *gin.Context) {
 		return
 	}
 
+	type_, err := keystore.GetFieldNameForKey(key)
```
I think we can simplify a bit by doing this once in the constructor: https://github.com/smartcontractkit/chainlink/pull/7514
WDYT?

jmank88 commented on core/services/chainlink/testdata/config-full.toml:
```diff
@@ -42,6 +42,13 @@ SendInterval = '1m0s'
+Enabled = false
+ForwardToUrl = 'http://localhost:9898'
+Environment = 'develop'
+JsonWrapperKey = ''
+Headers = ''
```
Oh, `JsonWrapperKey` and `Headers` still need example values.

jmank88 commented on core/config/presenters.go:
```diff
@@ -84,6 +84,9 @@ type EnvPrinter struct {
 	TriggerFallbackDBPollInterval              time.Duration   `json:"JOB_PIPELINE_DB_POLL_INTERVAL"`
 
+	// AuditLogger
+	AuditLoggerEnabled bool `json:"AUDIT_LOGGER_ENABLED"`
```
I think that all of them should be added here...did it only error for this one?

jmank88 replied:
You may still want to include the others here for completeness.

jmank88 commented on core/config/v2/types.go:
```diff
@@ -38,7 +39,8 @@ type Core struct {
 	TelemetryIngress *TelemetryIngress
 
-	Log *Log
+	AuditLogger *audit.AuditLoggerConfig
+	Log         *Log
```
```suggestion
	AuditLogger *audit.AuditLoggerConfig

	Log *Log
```

jmank88 commented on core/web/resolver/resolver_test.go:
```diff
@@ -109,8 +112,11 @@ func setupFramework(t *testing.T) *gqlTestFramework {
+		auditLogger: &audit.AuditLoggerService{},
 	}
 
+	app.Mock.On("GetAuditLogger", mock.Anything, mock.Anything).Return(&audit.AuditLoggerService{}).Maybe()
```
```suggestion
	app.Mock.On("GetAuditLogger", mock.Anything, mock.Anything).Return(audit.NopLogger).Maybe()
```

jmank88 commented on core/config/v2/docs/core.toml:
```diff
@@ -105,6 +105,16 @@ SendTimeout = '10s' # Default
+# ForwardToUrl is where you want to forward logs to
+ForwardToUrl = 'http://localhost:9898' # Example
+# JsonWrapperKey if set wraps the map of data under another single key to make parsing easier
+JsonWrapperKey = '' # Default
+# Headers is the set of headers you wish to pass along with each request
+Headers = '' # Default
```
Oh these should be `# Example` values in the case of zero value defaults.

jmank88 commented on core/services/chainlink/cfgtest/dump/full-custom.env:
```diff
@@ -33,6 +33,11 @@ DATABASE_LOCKING_MODE=advisory
+AUDIT_LOGGER_ENABLED=false
+AUDIT_LOGGER_FORWARD_TO_URL=http://localhost:9898
+AUDIT_LOGGER_JSON_WRAPPER_KEY=event
+AUDIT_LOGGER_HEADERS=
```
This one needs an example value to validate the dump method will convert properly.

jmank88 commented on core/cmd/local_client_test.go:
```diff
@@ -51,7 +52,7 @@ func TestClient_RunNodeShowsEnv(t *testing.T) {
 	db := pgtest.NewSqlxDB(t)
 	pCfg := cltest.NewTestGeneralConfig(t)
-	sessionORM := sessions.NewORM(db, time.Minute, lggr, pCfg)
+	sessionORM := sessions.NewORM(db, time.Minute, lggr, pCfg, &audit.AuditLoggerService{})
```
A couple of loose `&audit.AuditLoggerService{}` in here that could point to `audit.NoopLogger` instead.
